### ISSUE TITLE
macvlan maybe not fully correctly implemented

### ISSUE THREAD

**Issue by hadmut** [lxc/lxd#11319]
* Distribution: Ubuntu
 * Distribution version: 22.04
 * The output of "lxc info" see below

Hi, 
I am trying to use an LXD container (22.04) with a macvlan interface in order to be connected to the LAN and reachable from other machines. I'm using a profile, which is

config: {}
description: Anschluss an das LAN/Default-Interface enp4s0
devices:
  eth0:
    nictype: macvlan
    parent: enp4s0
    type: nic
name: macvlan
used_by:
- /1.0/instances/install
- /1.0/instances/better-parakeet

and indeed, the machine gets an IP adress from my LAN assigned by my internet router (and not from lxdbr0), so works as expected. 

But then, the virtual machine is not reachable, not ping-able from the host. I can ping it from other machines, but not from the host itself. I first thought that LXD is not setting the correct macvlan bridge mode, because in normal mode the device is not visible from the host, while in bridge mode it is (it resembles a simple bridge that doesn't need to learn). But the source code of LXD shows that it uses the bridge mode for macvlans. After doing some more research, I found this article

https://unix.stackexchange.com/questions/602384/why-i-cant-ping-between-two-siblings-macvlans-in-bridge-mode

where someone claims that macvlans do work correctly in bridge mode, when in different name spaces, because otherwise the kernel misinterpretes the ip packet and drops it. 

But ip netns  does not list any name spaces, neither on the host, nor on the guest. 

On the other hand, on 
https://discuss.linuxcontainers.org/t/running-containers-in-custom-network-namespace/12204
Stéphane Graber said that putting LXD containers into namespaces is not officially supported. 

So it seems that it doesn't work to use macvlans in bridge mode, just in the regular mode, even if the source code chooses the bridge mode. 

Somewhat confusing. 

lxc info: 
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- snapshot_schedule_aliases
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
- gpu_mdev
- resources_pci_iommu
- resources_network_usb
- resources_disk_address
- network_physical_ovn_ingress_mode
- network_ovn_dhcp
- network_physical_routes_anycast
- projects_limits_instances
- network_state_vlan
- instance_nic_bridged_port_isolation
- instance_bulk_state_change
- network_gvrp
- instance_pool_move
- gpu_sriov
- pci_device_type
- storage_volume_state
- network_acl
- migration_stateful
- disk_state_quota
- storage_ceph_features
- projects_compression
- projects_images_remote_cache_expiry
- certificate_project
- network_ovn_acl
- projects_images_auto_update
- projects_restricted_cluster_target
- images_default_architecture
- network_ovn_acl_defaults
- gpu_mig
- project_usage
- network_bridge_acl
- warnings
- projects_restricted_backups_and_snapshots
- clustering_join_token
- clustering_description
- server_trusted_proxy
- clustering_update_cert
- storage_api_project
- server_instance_driver_operational
- server_supported_storage_drivers
- event_lifecycle_requestor_address
- resources_gpu_usb
- clustering_evacuation
- network_ovn_nat_address
- network_bgp
- network_forward
- custom_volume_refresh
- network_counters_errors_dropped
- metrics
- image_source_project
- clustering_config
- network_peer
- linux_sysctl
- network_dns
- ovn_nic_acceleration
- certificate_self_renewal
- instance_project_move
- storage_volume_project_move
- cloud_init
- network_dns_nat
- database_leader
- instance_all_projects
- clustering_groups
- ceph_rbd_du
- instance_get_full
- qemu_metrics
- gpu_mig_uuid
- event_project
- clustering_evacuation_live
- instance_allow_inconsistent_copy
- network_state_ovn
- storage_volume_api_filtering
- image_restrictions
- storage_zfs_export
- network_dns_records
- storage_zfs_reserve_space
- network_acl_log
- storage_zfs_blocksize
- metrics_cpu_seconds
- instance_snapshot_never
- certificate_token
- instance_nic_routed_neighbor_probe
- event_hub
- agent_nic_config
- projects_restricted_intercept
- metrics_authentication
- images_target_project
- cluster_migration_inconsistent_copy
- cluster_ovn_chassis
- container_syscall_intercept_sched_setscheduler
- storage_lvm_thinpool_metadata_size
- storage_volume_state_total
- instance_file_head
- instances_nic_host_name
- image_copy_profile
- container_syscall_intercept_sysinfo
- clustering_evacuation_mode
- resources_pci_vpd
- qemu_raw_conf
- storage_cephfs_fscache
- network_load_balancer
- vsock_api
- instance_ready_state
- network_bgp_holdtime
- storage_volumes_all_projects
- metrics_memory_oom_total
- storage_buckets
- storage_buckets_create_credentials
- metrics_cpu_effective_total
- projects_networks_restricted_access
- storage_buckets_local
- loki
- acme
- internal_metrics
- cluster_join_token_expiry
- remote_token_expiry
- init_preseed
- storage_volumes_created_at
- cpu_hotplug
- projects_networks_zones
- network_txqueuelen
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIICETCCAZagAwIBAgIRAJDMIYKh2mB5uMA5R3bVaoowCgYIKoZIzj0EAwMwODEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEB3YXNj
    aGJyZXR0MB4XDTIwMTAyMjIwMjEwNloXDTMwMTAyMDIwMjEwNlowODEcMBoGA1UE
    ChMTbGludXhjb250YWluZXJzLm9yZzEYMBYGA1UEAwwPcm9vdEB3YXNjaGJyZXR0
    MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAEz+pnLLVYMaM/GmLBVx2v8oetgmvqZF4U
    gXlcq78ZXccW0dprz77cY9QSYZbyytc/uFDmQW96O7qqZRzaLj48HhX5RN4UwS6a
    4Keqb9UDAkJehpmF5txGQxECv3M9pOtIo2QwYjAOBgNVHQ8BAf8EBAMCBaAwEwYD
    VR0lBAwwCgYIKwYBBQUHAwEwDAYDVR0TAQH/BAIwADAtBgNVHREEJjAkggp3YXNj
    aGJyZXR0hwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqGSM49BAMDA2kAMGYC
    MQCh0vHntiQZCjV27KQcJBMgs8qBMh9fnOQVwDaqkXJHPHncTHsF8GcJ8v45QpdG
    fHUCMQD5sGPTeJN7xfkg4rGQVoGIMpV53760k+qXmTlzfaO7pKPtSFr6gmhQLb02
    uZAfi4w=
    -----END CERTIFICATE-----
  certificate_fingerprint: 75137d9d01638e1260ceff1829ec7f2c99c9dd3423b0783f7573658630c42363
  driver: lxc | qemu
  driver_version: 5.0.2 | 7.1.0
  firewall: nftables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    idmapped_mounts: "true"
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.15.0-58-generic
  lxc_features:
    cgroup2: "true"
    core_scheduling: "true"
    devpts_fd: "true"
    idmapped_mounts_v2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Ubuntu
  os_version: "22.04"
  project: default
  server: lxd
  server_clustered: false
  server_event_mode: full-mesh
  server_name: waschbrett
  server_pid: 6014
  server_version: "5.10"
  storage: dir
  storage_version: "1"
  storage_supported_drivers:
  - name: lvm
    version: 2.03.07(2) (2019-11-30) / 1.02.167 (2019-11-30) / 4.45.0
    remote: false
  - name: zfs
    version: 2.1.4-0ubuntu0.1
    remote: false
  - name: btrfs
    version: 5.4.1
    remote: false
  - name: ceph
    version: 15.2.17
    remote: true
  - name: cephfs
    version: 15.2.17
    remote: true
  - name: cephobject
    version: 15.2.17
    remote: true
  - name: dir
    version: "1"
    remote: false

**Comment by stgraber** [comment]
Unless things changed significantly since I last looked into this, `macvlan` allows for communication between the `macvlan` device and other `macvlan` devices on the system (when in bridge mode) as well as communication to outside of the system.

However it does not allow for hairpin traffic, that is, the primary network interface cannot talk to the macvlan devices, nor can the macvlan devices talk to the primary devices. That's not 100% true though as some physical switches can do hairpin for you, bouncing back your own port traffic back towards you to hit the host, but this isn't common.

The most common way to avoid this issue, other than using a bridge rather than macvlan devices is to have the host itself also use macvlan. In that scenario, the physical NIC is left unconfigured and all instances and host use macvlan for their networking, making it all work.

**Comment by stgraber** [comment]
Of the macvlan modes I know of, the behavior is:
 - private: macvlan to outside world only, will filter out hairpin traffic from external switch too
 - vepa: all traffic sent through external switch, does not filter hairpin from external
 - bridge: virtual built-in switch (bridge) that allows for forwarding between macvlan devices without external switch supporting hairpin
 - passthru: not really your normal macvlan. Only one macvlan of that type can be created and it effectively clones the physical device. It's basically an alternative to moving the physical device to the instance's namespace or through VFIO to a VM.

Note that all of those always refer to traffic between macvlan instances, in general it's assumed that the physical device isn't configured (especially required for passthru). In practice, most LXD users do have a configuring physical device, which usually works fine (minus some buggy drivers) but does come with the consequence of not being able to communicate with the instances running on the system.

**Comment by hadmut** [comment]
I'm currently following the suspicion, that this is not a bug of LXD itself, but of iproute2/kernel, since I found in LXD's source code that it makes use of the ip command, instead of direct kernel apis, and found similar problems when using ip directly. 

Let's suspend this issue until I got some reply. Unfortunately it is very difficult to get a reply from kernel developers and cumbersome to launch a bug report.

**Comment by hadmut** [comment]
I got somewhat further with debugging. 

The main problem is poor documentation of the macvlan driver, not even docs.kernel.org does have a proper documentation. Best I could find was from redhat:

https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#macvlan

As far as I can determine, the "bridge" mode and the thin description available is somewhat misleading, as it does *not*, opposed to what the description suggests, allow the host on its regular interface to communicate with a macvlan interface stacked on its interface and assigned to, e.g. an LXD, docker, podman container. It just enables communication between macvlan-devices stacked onto the same physical interfaces. 

So if the host itself wants to communicate with a container using a macvlan device, it needs – to the best of my current unreliable knowledge – to use a macvlan device (stacked onto its own physical device) itself in order to participate in the bridge mode. 

This is where it gets really nasty when using Ubuntu,  since Ubuntu uses either NetworkManager (desktops) or netplan (servers). It is not yet implemented in netplan, but netplan still occopies network configuration, thus making it tricky to workaround netplan. 

It gets really dirty if there is a further requirement, i.e. using a router (DSL, docsis, whatever) with the ability to wakeup devices using the wakeonlan. Because then the running computer (i.e. the host using its macvlan interface) must have the same MAC address as the unpowered computer. Although it is technically possible to assign the same MAC address to the physical and macvlan device, netplan and systemd-networkd and its tools cannot cope with that and remain hanging in the network configuration while bootup. 

So while I found the answer to the problem in theory, Ubuntu's half-cooked network configuration system (systemd+netplan) makes it really difficult and error prone to get this up and running, not to mention the lack of documentation. 

Although I may have understood what's going on to a certain degree, I do not consider this as production ready. 

It is, although not the fault of LXD, still a problem adjacent to LXD, since LXD is predestined to run virtual servers using macvlan to appear like physical servers in the network. This runs more or less out of the box, as long as the host itself does not need to communicate with the virtual container. 

BTW, to check proper configuration of a network device of both the type and the mode, use ip -d link  to show these details.  ip link without -d doesn't show.
